### DISCUSSION TITLE
Which calls should be server-server and which calls should be server-client?

### DISCUSSION THREAD

**Discussion by lennartvanlaake** [OpenHospitalityNetwork/fedi-trustroots#78] [Ideas]
Should the home-instance of a user proxy all requests of that user, or should eventually the user's browser get to send all the requests?

If it would be the browser, a home server can still add signed data to the user's session for calls to other servers if we would go for a JWT(like) solution, as described here: https://docs.google.com/document/d/1m-xz7gackRVwHEJvy07Y__SRKd87_z6zF8xoNwQ1mkc/edit

**Comment by chagai95** [comment]
I don't fully understand the advantages and disadvantages of this so I would appreciate some kind of explanation or even (messy) notes on what those might be.

Having said that I generally take the approach of just trying and seeing what I can get to work faster so for me that would be client to servers because I (with Mariha's help 😉)  found that in the code and it seems like I could manage to write a config that can then be used to be used to query each server independently, this is off course just as a proof of concept which might also be used to allow at least a few instances to be connected until we figure out more complex scenarios.

**Reply by lennartvanlaake** [reply]
I can make an attempt at a (biassed) explanation of the basic scenarios. Sorry if this is mostly about authentication, but that is the main point of difference between those two strategies.

_Admin from instance A wants to trust instance B_

I would say this would require instance A to do a call to instance B (server-server) to fetch the public ckey of instance B (and perhaps a name/logo/whatever to identify instance B). I would expect the public key and the name of the server to be publically available, so for this call no auth would be needed.

_User from instance A wants to interact (search/message/request) with instance B_

If we go for the client-server option, instance A would give a signed cookie (see JWT doc) that allows instance B to recognize that the user is indeed logged in to instance A. 

If we go for a server-server option, I can think of two options.

1. Server A would get a blanket authorisation to interact with server B. A user would send a request to server A to search server B, and server A would use its trust with server B to authorize this action and send the result back to the user.  This would be easier to implement, but you would not be able to distinguish between users in your permissions. So either all users of platform A gets to see/do something or none of them do.

2. Server A would add data of the current user session to its call with server B, together with some sort of token/whitelist system. I think this would be a pain to implement since this is a somewhat unusual way of adding session data to calls.

**Pro's of server-client communication**

- Easier to implement the authentication code since this is the most direct way of communicating 
- No extra work needed for implementing features: after authentication it is just a normal API call from a browser to a server.

**Cons of server-client communication**

- IP address is exposed to non-home instances
- ... ? (Maybe the server-server fans can add stuff here?


**Pro's of server-server communication**

- No IP address of user exposed since the home server acts as a proxy for all the user's requests
- ....?

**Cons of server-server communication**

- The code for doing something on your own instance and proxying calls to another platform would be different. This would mean a significant amount of extra work per feature.
- Proxying calls would be a less usual setup, so it would involve a good bit of thinking and custom solution finding to do that well
- Being able to differentiate between users in authorization would be a bit harder
- Extra processing pressure per instance

**Reply by weex** [reply]
Love the pros and cons approach. Would be awesome in a wiki and I'd suggest renaming server-client to "one client, many server communication".

Cons of server-client communication:
* Client load scales linearly with number of connected servers
* Server load scales linearly with total number of users

Do servers need to communicate to determine JWT validity or do they use their local CA stores as in https? I'll have to read more about JWT.

**Comment by weex** [comment]
Took a look at the google doc and I'm wary of commenting on that since it's ephemeral and later folks might not be able to access to see what all went into our decision(s). Can one of you summarize what's in that doc so it's here in the discussion record?

**Reply by lennartvanlaake** [reply]
This is the original document. I've ommitted Changai's edit for now, since it is more of a comment on the proposal rather than the proposal itself. If we want to use this Discussion as a source of reference, I think Changai's comment is probably best posted by him as part of this Discussion

> Why do we need this?
> 
> We want a user that has a profile on one instance, his “home” instance, to interact with other instances. It would therefore be useful if the session from the “home” instance can be verified and used by the other instances. 
> 
> How can we do this?
> 
> The technology behind this would be JSON Web Tokens (JWT). A JWT contains a signature based on the private key of the issuing instance (in our case the “home” server of the user) and a JSON object containing arbitrary data (probably a user name, the permissions of that user and the id/name of his home server) . Because of the signature included the integrity of the data in the JSON object can be verified using the corresponding public key of the issuing instance
> 
> How would this work in general?
> 
> A user would log into his “home” instance A
> The user gets a cookie containing the session signed by the private key of instance A
> The user makes a call to a second instance B with the cookie from instance A
> Instance B verifies the signature of the session with the public key of instance A
> Instance B now “knows” that this user has a valid session from instance A and can give him access accordingly
> The user can now make calls to instance B
> 
> How could we make this work for us?
> 
> This is just some random thoughts I have, definitely open for discussion
> 
> The public key of each instance does not have to be secret, it can be hosted publically on some URL. We could use this to have a very dynamic approach to a “whitelist” of other instances that are to be trusted. For example:
> 
> Instance A has a whitelist of URLs of public keys linked to instance names/ids that it trusts
> This list contains instance B with public key url X
> Instance A receives an API call from a user with a session that claims to from instance B
> Instance A looks up the public key url of instance B (url X)
> Instance A can now check based on the public key of instance B if the API call is actually from a user that has a valid session from instance B
> 
> How hard would this be?
> 
> Cryptographic stuff and auth is always a bit of a pain to do well. The entire application would need to have some sort of access to the active session to check if a “foreign” user is allowed to do a certain thing. 
> 
> The actual validation of JWT signatures does not look that complicated. I found a library here that seems very straight-forward, maintained by Auth0: https://github.com/auth0/node-jsonwebtoken
>
